### PR TITLE
Enumeration of Hot Rod protocol versions should be an Enum

### PR DESCRIPTION
**Pull request by rhusar**
Stumbled upon this while integrating into WildFly. The user-specified protocol version needs to be validated, currently there is no available list of possibly versions. Looks like the idea may have started in org.infinispan.client.hotrod.Version but was never expanded and is out of date.

### CODE REVIEW

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/Configuration.java [hunk 1]
```diff
@@ -157,7 +158,7 @@ public NearCacheConfiguration nearCache() {
       return nearCache;
    }
 
-   public String protocolVersion() {
+   public ProtocolVersion protocolVersion() {
```
Comment (tristantarrant) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/Configuration.java]: Let's not change this signature. Maybe add a separate method

Comment (rhusar) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/Configuration.java]: Ok, will do, so this is considered a public API thus should not change.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/ProtocolVersion.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package org.infinispan.client.hotrod;
+
+import java.util.EnumSet;
+import java.util.Locale;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
```
Comment (slaskawi) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/ProtocolVersion.java]: Could you please add some meaningful javadoc?

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/ProtocolVersion.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+   ProtocolVersion(int major, int minor) {
+      version = String.format(Locale.ROOT, "%d.%d", major, minor);
+   }
+
+   @Override
+   public String toString() {
+      return version;
+   }
+
+   private static final Map<String, ProtocolVersion> versions = EnumSet.allOf(ProtocolVersion.class).stream().collect(Collectors.toMap(ProtocolVersion::toString, Function.identity()));
```
Comment (slaskawi) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/ProtocolVersion.java]: Could you please move it up, where the other fields are placed?

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AbstractConfigurationChildBuilder.java [hunk 4]
```diff
@@ -105,9 +106,15 @@ public ConfigurationBuilder marshaller(Marshaller marshaller) {
       return builder.marshaller(marshaller);
    }
 
+   @Deprecated
```
Comment (slaskawi) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/configuration/AbstractConfigurationChildBuilder.java]: Please add some javadoc and information which method should be used instead.

**Comment by slaskawi** [comment]
Nice work! Only some picky stuff found..

**Comment by wburns** [comment]
Changes seem good overall to me as well.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/ProtocolVersion.java (hunk 2 above)
Comment (rhusar) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/ProtocolVersion.java]: Done.

**Comment by tristantarrant** [comment]
I added a commit on my branch, @rhusar please integrate them here https://github.com/tristantarrant/infinispan/commit/a38430926bd67c2a0bd21d5c8172e26fdb8d8d66

**Comment by rhusar** [comment]
Pulled the commit here.

**Comment by tristantarrant** [comment]
Assigned a Jira, fixed one last issue, squashed and merged. Thanks

**Comment by rhusar** [comment]
Thanks Tristan.

**Pull request closed without merging**